source/SkiaSharp.Views.Forms/SkiaSharp.Views.Forms.Native.Shared/SKCanvasViewRendererBase.cs

```diff
@@ -0,0 +1,129 @@
+using SKNativePaintSurfaceEventArgs = SkiaSharp.Views.UWP.SKPaintSurfaceEventArgs;
+#endif
+
+namespace SkiaSharp.Views.Forms
+{
+	public abstract class SKCanvasViewRendererBase<TFormsView, TNativeView> : ViewRenderer<TFormsView, TNativeView>
+		where TFormsView : SKFormsView
+		where TNativeView : SKNativeView
+	{
+		protected override void OnElementChanged(ElementChangedEventArgs<TFormsView> e)
+		{
+			if (e.OldElement != null)
+			{
+				var oldController = (ISKCanvasViewController)e.OldElement;
+
+				// unsubscribe from events
+				oldController.SurfaceInvalidated -= OnSurfaceInvalidated;
+				oldController.GetCanvasSize -= OnGetCanvasSize;
+			}
+
+			if (Control != null)
+			{
+				var control = Control;
+				control.PaintSurface -= OnPaintSurface;
+			}
+
+			if (e.NewElement != null)
+			{
+				var newController = (ISKCanvasViewController)e.NewElement;
+
+				// create the native view
+				var view = CreateNativeControl();
+				view.IgnorePixelScaling = e.NewElement.IgnorePixelScaling;
+				view.PaintSurface += OnPaintSurface;
+				SetNativeControl(view);
+
+				// subscribe to events from the user
+				newController.SurfaceInvalidated += OnSurfaceInvalidated;
+				newController.GetCanvasSize += OnGetCanvasSize;
+
+				// paint for the first time
+				OnSurfaceInvalidated(newController, EventArgs.Empty);
+			}
+
+			base.OnElementChanged(e);
+		}
+		
+#if __ANDROID__
+		protected override TNativeView CreateNativeControl()
+		{
+			return (TNativeView)Activator.CreateInstance(typeof(TNativeView), new[] { Context });
+		}
+#else
+		protected virtual TNativeView CreateNativeControl()
+		{
+			return (TNativeView)Activator.CreateInstance(typeof(TNativeView));
+		}
+#endif
+
+		protected override void OnElementPropertyChanged(object sender, PropertyChangedEventArgs e)
+		{
+			base.OnElementPropertyChanged(sender, e);
+
+			if (e.PropertyName == nameof(SKFormsView.IgnorePixelScaling))
+			{
+				Control.IgnorePixelScaling = Element.IgnorePixelScaling;
+			}
+		}
+
+		protected override void Dispose(bool disposing)
+		{
+			// detach all events before disposing
+			var controller = (ISKCanvasViewController)Element;
+			if (controller != null)
+			{
+				controller.SurfaceInvalidated -= OnSurfaceInvalidated;
+				controller.GetCanvasSize -= OnGetCanvasSize;
+			}
+
+			var control = Control;
+			if (control != null)
+			{
+				control.PaintSurface -= OnPaintSurface;
+			}
+
+			base.Dispose(disposing);
+		}
+
+		private void OnPaintSurface(object sender, SKNativePaintSurfaceEventArgs e)
+		{
+			var controller = Element as ISKCanvasViewController;
+
+			// the control is being repainted, let the user know
+			controller?.OnPaintSurface(new SKPaintSurfaceEventArgs(e.Surface, e.Info));
+		}
+
+		private void OnSurfaceInvalidated(object sender, EventArgs eventArgs)
+		{
+			// repaint the native control
+#if __IOS__
+			Control.SetNeedsDisplay();
+#else
+			Control.Invalidate();
+#endif
+		}
```

Comment:
This one is a bit tough... I think the two sides of the fence is to either let the user make sure that invalidate is done from the main/UI thread via `Xamarin.Forms.Device.BeginInvokeOnMainThread()`, or to let the framework do it automatically.

For most cases, the app will always be using the main thread, so this should not be an issue. Thus, the overhead for posting to the main thread on each call might be undesirable. However, if some event out of the user's control is causing this to be called, then we will have to wrap the invalidate.

In the case of your beta, are you able to determine if a call was being made from a background thread that you could control?